src/compiler/program.ts

```diff
@@ -2667,6 +2667,11 @@ namespace ts {
                 return isSameFile(filePath, out) || isSameFile(filePath, removeFileExtension(out) + Extension.Dts);
             }
 
+            // If declarationDir is specified, return if its a file in that directory
+            if (options.declarationDir && containsPath(options.declarationDir, filePath, currentDirectory, !host.useCaseSensitiveFileNames())) {

```

Comment:
should we also check the file extension?

Reply:
When "files" is not provided in the tsconfig, declarationDir is added to excludes. We dont check extensions in there so if we go by same rule, we needn't check extension?
Also in outDir, we dont see if that is really output file or some random file.